### PR TITLE
Enable propagation of the trace context from collector

### PR DESCRIPTION
**Pull request by pavankrish123**
**Description:** 

Added support to propagate traceparent etc headers from collector to the downstream entities that process the export requests from collector. 

**Link to tracking Issue:** [5572](https://github.com/open-telemetry/opentelemetry-collector/issues/5572) 

**Testing:** 

running 

```bash
./otelcorecol_darwin_amd64 --config test.yaml 

```
gives 

```bash 
2022/07/27 21:59:24 request headers: map[Accept-Encoding:[gzip] B3:[e027c66374cdfd8c54704a92cd7759eb-7ce684e76639b14e-0] Content-Encoding:[gzip] Content-Length:[860] Content-Type:[application/x-protobuf] Traceparent:[00-e027c66374cdfd8c54704a92cd7759eb-7ce684e76639b14e-00] User-Agent:[Local OpenTelemetry Collector binary, testing only./0.55.0-dev (darwin/amd64)]]
```
for config 

``` yaml 
receivers:
  otlp:
    protocols:
      grpc:

exporters:
  otlphttp:
    endpoint: http://0.0.0.0:3000
    traces_endpoint: http://0.0.0.0:3000/v1/trace
    tls:
      insecure: true

  logging:


processors:
  batch:

service:
  telemetry:
    traces:
      propagators:
        - tracecontext
        - b3
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [otlphttp, logging]
      processors: [batch]
```

(and)

```bash 
2022/07/28 00:14:57 request headers: map[Accept-Encoding:[gzip] Content-Encoding:[gzip] Content-Length:[859] Content-Type:[application/x-protobuf] User-Agent:[Local OpenTelemetry Collector binary, testing only./0.55.0-dev (darwin/amd64)]]
```
for config 
```
receivers:
  otlp:
    protocols:
      grpc:

exporters:
  otlphttp:
    endpoint: http://0.0.0.0:3000
    traces_endpoint: http://0.0.0.0:3000/v1/trace
    tls:
      insecure: true

  logging:


processors:
  batch:

service:
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [otlphttp, logging]
      processors: [batch]
```

### CODE REVIEW

File: service/telemetry/config.go
Comment (mx-psi) [inline service/telemetry/config.go:110]: We should have a more general description here
```suggestion
// TracesConfig defines the configurable settings for service telemetry traces.
```

The other details are implementation details that I don't think we should document, since it may get out of sync if we decide to use a different propagator.

File: service/telemetry/config.go
Comment (mx-psi) [inline service/telemetry/config.go:114]: I think we should have a propagation section with an enabled field, since (i) this is the pattern we follow elsewhere and (ii) it's possible that we will get other configuration settings related to propagation. So, something like:

```
traces:
  propagator:
    enabled: true
```

I am not familiar with propagation, so maybe the approach should be different (should you have to choose what propagator to use?)

File: service/telemetry/config.go
Comment (mx-psi) [inline service/telemetry/config.go:113]: We should have it off by default for now, and think about having it on by default later, after it has had some testing.

File: cmd/otelcorecol/go.mod
Comment (mx-psi) [inline cmd/otelcorecol/go.mod:62]: IIUC this is a direct dependency, right? So it should be on the other require block

**Review by mx-psi — COMMENTED** [review]
👋 I think you should open an issue to discuss the use case for this before we decide on this PR.

I also left some suggestions to make the configuration more consistent with existing practices on the Collector.

File: service/telemetry/config.go
Comment (pavankrish123) [reply service/telemetry/config.go:113]: @jpkrohling let me know how to proceed. Per your comment in https://github.com/open-telemetry/opentelemetry-collector/issues/5572 I have enabled it by default, shall I heed @mx-psi advice here?

File: service/telemetry/config.go
Comment (pavankrish123) [reply service/telemetry/config.go:112]: I like this structure propagation does have other possible configurations - like (type of propagator etc) - Let me restructure the propagation into its own nested struct and reissue a PR.

File: cmd/otelcorecol/go.mod
Comment (pavankrish123) [reply cmd/otelcorecol/go.mod:62]: Think this is In the context of cmd/otelcorecol module - the files in the module doesn't refer to autoprop directly (just as they don't refer to other indirect dependencies, even though they are direct dependencies for exporters, receivers, service and extensions etc. 

The top level go.mod present in the root folder has go.opentelemetry.io/contrib/propagators/autoprop as direct dependency though.

**Comment by pavankrish123** [comment]
> 👋 I think you should open an issue to discuss the use case for this before we decide on this PR.
@mx-psi the linked issue is in the PR description.  https://github.com/open-telemetry/opentelemetry-collector/issues/5572

File: cmd/otelcorecol/go.mod
Comment (mx-psi) [reply cmd/otelcorecol/go.mod:62]: Right :facepalm: looking at the wrong file, thanks!

File: service/telemetry/config.go
Comment (jpkrohling) [reply service/telemetry/config.go]: In the long term, it should be enabled by default, so that we won't break the chain. Given that this is a change in the current behavior, this should be a two-step change: implement the new feature and make it disabled by default, but leave a note to the changelog stating that this will be the enabled by default by N+2.

Comment (pavankrish123) [reply service/telemetry/config.go]: Thanks @jpkrohling - I have set the propagation to false by default for now - @mx-psi please review and let me know.

**Review by mx-psi — APPROVED** [review]
The config struct looks sensible, and it sounds like a feature we can support at this level. 

I don't feel very comfortable with comitting to autoprop which has had a single version so far and is not 1.0, but I guess since we can still make breaking changes we can change our mind later (specially if we have this off-by-default for now)

File: service/telemetry/config.go
Comment (Aneurysm9) [reply service/telemetry/config.go]: Should this even have an enable/disable flag in the config, or should it be governed by a feature gate until we're comfortable enabling it by default?  The user can always use the `none` "propagator" to disable context propagation.

File: service/telemetry/config.go
Comment (Aneurysm9) [inline service/telemetry/config.go:118]: See [my comment](https://github.com/open-telemetry/opentelemetry-collector/pull/5670#discussion_r918731880) elsewhere.  I don't think `enabled` is the right option here.  I'm not sure any option is required given that propagation can be disabled by using the `none` "propagator".

File: service/service.go
Comment (Aneurysm9) [inline service/service.go:67]: If this is a feature gate instead we can eliminate this once we're comfortable having propagation enabled by default.  The `none` "propagator" still provides a way for users to prevent propagation.

Comment (pavankrish123) [reply service/service.go:67]: @Aneurysm9 just for the sake of my understanding - some clarifications (sorry never dealt with feature gate module before) 
1. By feature gate are you referring to https://github.com/open-telemetry/opentelemetry-collector/tree/main/service/featuregate? 
2. If that is what you are referring to, are you suggesting that we enable the context propagation by default (assuming this is what we want to do eventually) behind a feature gate flag and users can choose to set the OTEL_PROPAGATOR to `none` should they choose to disable the propagation. Thus eliminating need to have any configuration defined completely? 
3. Is there a good example on some features behind the feature gates?

Comment (mx-psi) [reply service/service.go:67]: The feature gate sounds good to me and allows us to commit less strongly to this, but I don't think `OTEL_PROPAGATOR` should be the only way to configure this when this is stable (to be honest, I think it's bad that it can be configured at all through an environment variable that we don't control, but I guess we can't avoid that unless the OTel Go SDK changes). If `enabled` is not appropriate, we should have a configuration option equivalent to `OTEL_PROPAGATOR` on the YAML.

**Comment by pavankrish123** [comment]
@Aneurysm9 adjusted the PR per your suggestions. Please let me know WDYT.

File: service/telemetry/config.go
Comment (pavankrish123) [reply service/telemetry/config.go]: per suggestion from @Aneurysm9 - the propagation is enabled by default behind a feature flag :)

**Comment by pavankrish123** [comment]
@jpkrohling please review the PR and let me know.

File: service/service.go
Comment (Aneurysm9) [reply service/service.go]: > we should have a configuration option equivalent to `OTEL_PROPAGATOR` on the YAML.

Agreed.  I think this could be accomplished currently by using `os.SetEnv` to set the value of `OTEL_PROPAGATOR` if it isn't already set.  I'll propose that an option be added to the autoprop interface to allow setting a default propagator by ID rather than just by providing an implementation.

> (to be honest, I think it's bad that it can be configured at all through an environment variable that we don't control, but I guess we can't avoid that unless the OTel Go library changes)

This is not going to change.  OTel in general has made the decision that SDKs should be configurable through environment variables and this is in line with that policy.  I think there is significant benefit to operators that may not be able to alter configuration in other ways but do have the ability to control the execution environment.  Ultimately, it would be little different from this configuration:

```yaml
service:
  telemetry:
    trace:
      propagator: ${OTEL_PROPAGATOR}
```

File: service/telemetry.go
Comment (Aneurysm9) [inline service/telemetry.go:140]:
```suggestion
		if _, exists := os.LookupEnv("OTEL_PROPAGATOR); !exists && cfg.Traces.Propagator != "" {
			// handle the error, though
			_  = os.SetEnv("OTEL_PROPAGATOR", cfg.Traces.Propagator)
		}
		otel.SetTextMapPropagator(autoprop.NewTextMapPropagator())
```

Something like this could potentially be used to enable control of the selector propagators via config if it isn't already set via environment.

Comment (pavankrish123) [reply service/telemetry.go:140]: @Aneurysm9 can we provide a factory method to construct the right propagators from comma separated strings in the autoprop module -(which I think is the right place). Setting environment variables from within the process doesn't seem clean and side effect-y

IMHO env variables is something that operators should drive. I suggest we document OTEL_PROPAGATOR and introduce configuration `cfg.tracer.Propgators` only when auto prop adds the factory method. Besides having the factory method take in the same input as what gets passed to OTEL_PROPAGATOR makes the module more consistent. 

cc: @mx-psi  @jpkrohling

File: service/service.go
Comment (pavankrish123) [reply service/service.go]: >>>  I'll propose that an option be added to the autoprop interface to allow setting a default propagator by ID rather than just by providing an implementation.

+1

File: service/telemetry.go
Comment (pavankrish123) [reply service/telemetry.go:143]: Had a chat with @Aneurysm9 - We have decided to leave the code as is with autoprop and will let users configure it via OTEL_PROPAGATORS. Once the id based factory api is available, will intoduce collector telemetry.traces.propagators configuration.

**Review by mx-psi — APPROVED** [review]

File: service/telemetry.go
Comment (jpkrohling) [inline service/telemetry.go:139]: This absolutely needs to be documented. Can it be done as part of this PR?

Comment (pavankrish123) [reply service/telemetry.go:139]: Thanks @jpkrohling   I will add the section to telemetry doc in another PR after this is merged.

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
As a rule of thumb in the collector we are moving away from env variables in general (allow them only as part of the yaml config via embedding). So I think we should offer an yaml alternative first for this feature.

**Comment by jpkrohling** [comment]
> As a rule of thumb in the collector we are moving away from env variables in general

While I agree with the general feeling, this goes into a bigger standard: OpenTelemetry SDK should generally be configurable via env vars.

**Comment by bogdandrutu** [comment]
> OpenTelemetry SDK should generally be configurable via env vars.

Unfortunately this is true because the TC members were lazy (starting with the person who writes this) and have not put the effort into review/propose a configuration (yaml/json) and lazily accepted env vars as the only way. /cc @jsuereth

**Comment by pavankrish123** [comment]
@bogdandrutu, I have updated PR per your suggestion using the new API added to `autoprop` module in `otel-go-contrib`. The API is not released yet, so I will keep the PR in draft. But please take a look into the new commit and let me know if you think I am heading in the right direction, I am adding unit tests  as well. 

Latest tests - 


running 

```bash
./otelcorecol_darwin_amd64 --config test.yaml --feature-gates=telemetry.allowTraceContextPropagation
```

gives 

```bash 
2022/07/27 21:59:24 request headers: map[Accept-Encoding:[gzip] B3:[e027c66374cdfd8c54704a92cd7759eb-7ce684e76639b14e-0] Content-Encoding:[gzip] Content-Length:[860] Content-Type:[application/x-protobuf] Traceparent:[00-e027c66374cdfd8c54704a92cd7759eb-7ce684e76639b14e-00] User-Agent:[Local OpenTelemetry Collector binary, testing only./0.55.0-dev (darwin/amd64)]]
```
for config 

``` yaml 
receivers:
  otlp:
    protocols:
      grpc:

exporters:
  otlphttp:
    endpoint: http://0.0.0.0:3000
    traces_endpoint: http://0.0.0.0:3000/v1/trace
    tls:
      insecure: true

  logging:


processors:
  batch:

service:
  telemetry:
    traces:
      propagators: "tracecontext,b3"
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [otlphttp, logging]
      processors: [batch]
```

(and)

```bash 
2022/07/28 00:14:57 request headers: map[Accept-Encoding:[gzip] Content-Encoding:[gzip] Content-Length:[859] Content-Type:[application/x-protobuf] Traceparent:[00-62f74abcde1f2ca18f0bb7faa8ba385a-9cc603f55b09d95f-00] User-Agent:[Local OpenTelemetry Collector binary, testing only./0.55.0-dev (darwin/amd64)]]
```
for config 
```
receivers:
  otlp:
    protocols:
      grpc:

exporters:
  otlphttp:
    endpoint: http://0.0.0.0:3000
    traces_endpoint: http://0.0.0.0:3000/v1/trace
    tls:
      insecure: true

  logging:


processors:
  batch:

service:
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [otlphttp, logging]
      processors: [batch]
```

File: service/telemetry.go
Comment (jpkrohling) [inline service/telemetry.go:142]: This feature requires the use of OpenTelemetry SDK, which is part of another feature gate, right?

**Review by jpkrohling — COMMENTED** [review]
Docs or it didn't happen 😉

File: service/telemetry.go
Comment (pavankrish123) [reply service/telemetry.go:142]: Actually no, the feature is merged in master (not yet released) - no feature gates here.

Comment (pavankrish123) [reply service/telemetry.go:143]: @jpkrohling  - I will move the PR from draft when go contrib officially releases the feature - Docs will be included - Waiting for the release to happen.

Comment (pavankrish123) [reply service/telemetry.go:143]: @jpkrohling Actually no, that feature gate is for using OpenTelemetry SDK metrics (only) - OpenTelemetry SDK was always used for tracing without any feature flags.

**Comment by pavankrish123** [comment]
Hello @bogdandrutu @jpkrohling - The autoprop api is officially released and I have updated the PR. The PR is complete. Please review and give your blessings. :) 

- The feature is available as a collector configuration now per @bogdandrutu suggestion and it is behind a feature gate.
-  I will work on getting docs added under https://opentelemetry.io/docs/collector once the PR is merged. There is no documentation as such in this context (collectors self telemetry) - So will have to write one from scratch.

File: cmd/builder/go.mod
Comment (jpkrohling) [inline cmd/builder/go.mod:43]: What happened here?

File: go.mod
Comment (jpkrohling) [inline go.mod:91]: Was this auto-formatted?

**Review by jpkrohling — COMMENTED** [review]
Please, revert the "retract" changes to the go.mod files. If they were intentional, please open a new PR with those isolated changes.

Other than that, LGTM, but I believe @bogdandrutu needs to sign off as well, given he had comments before.

File: go.mod
Comment (pavankrish123) [reply go.mod]: yes, fixed now

File: cmd/builder/go.mod
Comment (pavankrish123) [reply cmd/builder/go.mod]: fixed now

**Comment by pavankrish123** [comment]
> Please, revert the "retract" changes to the go.mod files. If they were intentional, please open a new PR with those isolated changes.
> 
> Other than that, LGTM, but I believe @bogdandrutu needs to sign off as well, given he had comments before.

done!

**Comment by pavankrish123** [comment]
>>> As a rule of thumb in the collector we are moving away from env variables in general (allow them only as part of the yaml config via embedding). So I think we should offer an yaml alternative first for this feature.

(Done!)

@bogdandrutu can you please review and let me know.

File: service/telemetry/config.go
Comment (bogdandrutu) [inline service/telemetry/config.go:113]: Why not a proper list in yaml?

File: service/telemetry/config.go
Comment (bogdandrutu) [inline service/telemetry/config.go:27]: Do we need the config? can we rely only on the featuregate initially?

Comment (pavankrish123) [reply service/telemetry/config.go:27]: Hi @bogdandrutu, i think it is necessary unless we are ok enabling just a default propagator once the feature gate is passed. I was instructed to have a configurable option (by you :-) )   in collectors configurations in the last commit, so I added this change. 

let me know if you changed your mind, I can remove this and just support the feature gate + default propagator.

File: service/telemetry/config.go
Comment (bogdandrutu) [reply service/telemetry/config.go:113]: Also add a comment that this is experimental, until otel defines standard yaml configuration for the library.

File: service/telemetry/config.go
Comment (bogdandrutu) [reply service/telemetry/config.go:27]: I am just a bit worried that we are defining this before otel does, so let's just add a note that this is experimental until otel defines a yaml config.

File: cmd/otelcorecol/go.mod
Comment (bogdandrutu) [inline cmd/otelcorecol/go.mod:62]: Worried about this dependency, since may get quite big. Can we limit to a subset of propagators for the moment?

File: service/telemetry/config.go
Comment (pavankrish123) [reply service/telemetry/config.go]: Done! updated the PR

**Comment by pavankrish123** [comment]
Updated the PR per @bogdandrutu suggestion, the configuration is a list now 

```yaml
receivers:
  otlp:
    protocols:
      grpc:

exporters:
  otlphttp:
    endpoint: http://0.0.0.0:3000
    traces_endpoint: http://0.0.0.0:3000/v1/trace
    tls:
      insecure: true

  logging:


processors:
  batch:

service:
  telemetry:
    traces:
      propagators:
        - tracecontext
        - b3
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [otlphttp, logging]
      processors: [batch]
```

```bash
./otelcorecol_darwin_amd64 --config test.yaml --feature-gates=telemetry.allowTraceContextPropagation
```

produces the following headers in the export request. 

```bash
2022/08/12 11:13:30 request headers: map[Accept-Encoding:[gzip] B3:[d6550dfa8b9c5e0f10acd6533391f08f-4bd7dc181ad45baf-0] Content-Encoding:[gzip] Content-Length:[276] Content-Type:[application/x-protobuf] Traceparent:[00-d6550dfa8b9c5e0f10acd6533391f08f-4bd7dc181ad45baf-00] User-Agent:[Local OpenTelemetry Collector binary, testing only./0.58.0-dev (darwin/amd64)]]
```

File: cmd/otelcorecol/go.mod
Comment (pavankrish123) [reply cmd/otelcorecol/go.mod:62]: Good concern to have, I think b3 and traceparent are the most used ones. Let me see if I can compile without getting other ones.

File: service/telemetry.go
Comment (bogdandrutu) [reply service/telemetry.go:140]: I think a simpler implementation is to make this featuregate just control the default of the configuration. So the list is empty if the gate is not enabled, otherwise is `["tracecontext]`? What do you think? Then be default we don't send this, also we have the gate to change the behavior to be default send, and for people that enable the config works without any gate?

File: cmd/otelcorecol/go.mod
Comment (bogdandrutu) [reply cmd/otelcorecol/go.mod:62]: Waiting on this?

Comment (pavankrish123) [reply cmd/otelcorecol/go.mod:62]: @bogdandrutu the way to limit the propagtors is to not use autoprop module, the module  - if we go about using go-contrib/propagators directly then we can limit - The propagators can be enabled via configuration strings but the logic will become sort of parallel/redundant  to what autoprop is doing but with limited options. Are you ok with this ?

**Comment by pavankrish123** [comment]
@bogdandrutu - just summarizing the suggestions from you, to make  sure I didn;t miss any 

1.  Limiting the propagators that we allow (tracecontext, b3) - this can be done if we don't use autoprop module at all(which pulls all the propgators)
2. Changing the default behavior of the feature gate  
   1. when feature gate flag is passed and there is no configuration in collectors yaml, we enable tracecontext by default 
   2. when feature gate flag is not passed and there is no configration in collectors yaml, we don;t propagate context at all. 
   3. when `telemetry.traces` configuration is present, we ignore the feature gate flag and honor the configuration and propagate the context 

Please let me know if my understanding of these changes you suggested is correct?

**Comment by bogdandrutu** [comment]
> Limiting the propagators that we allow (tracecontext, b3) - this can be done if we don't use autoprop module at all(which pulls all the propgators)

Correct

>Changing the default behavior of the feature gate
when feature gate flag is passed and there is no configuration in collectors yaml, we enable tracecontext by default
when feature gate flag is not passed and there is no configration in collectors yaml, we don;t propagate context at all.
when telemetry.traces configuration is present, we ignore the feature gate flag and honor the configuration and propagate the context

This is too complicated, let's go with a simple thing. Remove the featuregate and make default config to be "empty list" which means this is disable by default. Then in a separate PR we can discuss about adding a featuregate to change the default config.

File: cmd/otelcorecol/go.mod
Comment (pavankrish123) [reply cmd/otelcorecol/go.mod]: Done!

**Comment by pavankrish123** [comment]
Thanks @bogdandrutu - updated the PR per your suggestions. Kindly review

File: CHANGELOG.md
Comment (bogdandrutu) [inline CHANGELOG.md:20]: Please update.

**Review by bogdandrutu — APPROVED** [review]

File: CHANGELOG.md
Comment (pavankrish123) [reply CHANGELOG.md]: oops! Fixed it - Please review and let me know WDYT?

Comment (pavankrish123) [reply CHANGELOG.md]: fixed the PR description  as well.

Comment (pavankrish123) [reply CHANGELOG.md]: thanks @bogdandrutu,  fixed the PR description as well.

**Pull request merged**
